File: features/miso/loop/cards/browse/map/map.js
const feature_Map = {
  // the Leaflet instance and its host, both made once and kept for the life
  // of the page. The host lives OUTSIDE #app so the loop's wholesale repaint
  // cannot take the map away — /keep's #cardRemove idiom, for the same reason
  // at a larger size: re-mounting a map on every event would refetch tiles,
  // lose the pan and flicker.
  host: null,
  map: null,
  layer: null,
  markers: [],
  sig: '',      // the pins we last drew, so a repaint that changes nothing does nothing
  fitted: false,
  located: false,
  ATTRIB: '© OpenStreetMap contributors',

  // ---- what the page is showing ----------------------------------------
  // #mapData is Rust's whole contribution to the screen: an empty element
  // carrying the located cards. Present means "the map view is up".

  sync() {
    const data = document.getElementById('mapData');
    if (!data) { this.hide(); return; }
    this.show();
    if (!this.mount()) return;
    let pins = [];
    try { pins = JSON.parse(data.getAttribute('data-pins') || '[]'); } catch (e) { pins = []; }
    this.draw(pins);
    // the host is display:none until now, so Leaflet measured nothing
    this.map.invalidateSize();
  },

  show() {
    if (!this.host || this.host.style.display === 'block') return;
    this.host.style.display = 'block';
  },

  hide() {
    if (!this.host || this.host.style.display === 'none') return;
    this.host.style.display = 'none';
  },

  // ---- the map, made once ----------------------------------------------

  mount() {
    if (this.map) return true;
    if (typeof L === 'undefined' || !this.host) return false;   // vendored file missing: grey, not broken
    this.map = L.map(this.host, {
      zoomControl: false,
      attributionControl: true,
      // the pins carry the tap; a double tap is a zoom, and a stray double
      // tap on a pin should not also open the card
      tapTolerance: 20,
    });
    this.map.setView([51.2719, 0.1904], 3);
    this.map.attributionControl.setPrefix('');
    this.layer = L.tileLayer('tiles/{z}/{x}/{y}.png', {
      maxZoom: 19,
      attribution: this.ATTRIB,
      // one zoom's worth of what is on screen and no more: OSM's policy
      // forbids bulk prefetch and our proxy pays for every miss
      keepBuffer: 1,
      updateWhenIdle: true,
    }).addTo(this.map);
    this.credit();
    return true;
  },

  // the credit belongs to whatever source the server is proxying, so it is
  // asked for rather than assumed. One request, at mount; if it fails the
  // OpenStreetMap line stands, which is true of every source we would use.
  credit() {
    fetch('tiles/attribution')
      .then((r) => (r.ok ? r.text() : ''))
      .then((t) => {
        const line = (t || '').trim();
        if (!line || !this.layer) return;
        this.ATTRIB = line;
        this.map.attributionControl.removeAttribution(
          '© OpenStreetMap contributors');
        this.map.attributionControl.addAttribution(line);
      })
      .catch(() => {});
  },

  // ---- the pins ----------------------------------------------------------

  draw(pins) {
    const sig = JSON.stringify(pins.map((p) => [p.id, p.lat, p.lon, p.title, (p.face || '').length]));
    if (sig === this.sig) return;
    this.sig = sig;
    for (const m of this.markers) m.remove();
    this.markers = [];
    const at = [];
    for (const p of pins) {
      if (typeof p.lat !== 'number' || typeof p.lon !== 'number') continue;
      const m = L.marker([p.lat, p.lon], {
        icon: L.divIcon({
          className: 'map-pin-icon',
          html: this.pinHtml(p),
          iconSize: [40, 50],
          iconAnchor: [20, 50],
        }),
        title: p.title || '',
        keyboard: false,
      }).addTo(this.map);
      // the pin's own handler rather than the loop's delegated [data-ev]
      // listener: Leaflet stops the DOM event on its own markers, so the tap
      // is sent by hand — to the SAME event /browse already answers.
      m.on('click', () => {
        if (typeof feature_Loop !== 'undefined') {
          feature_Loop.send({ type: 'click', ev: 'browse_open:' + p.id });
        }
      });
      this.markers.push(m);
      at.push([p.lat, p.lon]);
    }
    if (!at.length) { this.locate(); return; }
    // fit once, and again only when the set of pins itself changes — a refit
    // on every repaint would fight the hand that just dragged the map
    this.map.fitBounds(L.latLngBounds(at).pad(0.25), {
      animate: false,
      maxZoom: 16,
    });
    this.fitted = true;
  },

  pinHtml(p) {
    const inner = p.face
      ? '<img src="' + this.esc(p.face) + '" alt="">'
      : '<span>' + this.esc(p.initial || '') + '</span>';
    return '<div class="map-pin"><div class="map-pin-face">' + inner
      + '</div><div class="map-pin-stem"></div></div>';
  },

  esc(s) {
    return String(s).replace(/&/g, '&amp;').replace(/</g, '&lt;')
      .replace(/>/g, '&gt;').replace(/"/g, '&quot;');
  },

  // nothing has a place yet: centre on this device if it will say, once, and
  // otherwise leave the world where it is. No error either way — a map with
  // nothing on it is a map.
  locate() {
    if (this.fitted || this.located || !navigator.geolocation) return;
    this.located = true;
    navigator.geolocation.getCurrentPosition(
      (pos) => {
        if (this.fitted) return;
        this.map.setView([pos.coords.latitude, pos.coords.longitude], 14);
      },
      () => {},
      { timeout: 6000, maximumAge: 300000 });
  },
};

{
  // the host, made at load and living outside #app so a repaint cannot take
  // it away. Hidden until a render says the map view is up.
  const fm_mapHost = document.createElement('div');
  fm_mapHost.id = 'misoMap';
  fm_mapHost.style.display = 'none';
  document.body.appendChild(fm_mapHost);
  feature_Map.host = fm_mapHost;

  // /loop's paint seam, taken by replacing the property at load — /keep's
  // idiom, and NOT a timer-installed wrapper (notes.md, "the apply-wrapper
  // race"). The map is synced after the html is on the screen, because that
  // is when #mapData exists to be found.
  if (typeof feature_Loop !== 'undefined') {
    const fm_mapPaint = feature_Loop.paint;
    feature_Loop.paint = function (html) {
      fm_mapPaint.call(this, html);
      try {
        feature_Map.sync();
      } catch (e) {
      }
    };
  }

  // a rotated phone, or a keyboard closing: Leaflet must be told the box
  // changed size, and only while it is on screen.
  window.addEventListener('resize', () => {
    if (feature_Map.map && feature_Map.host.style.display === 'block') {
      feature_Map.map.invalidateSize();
    }
  });
}

File: features/miso/loop/cards/browse/map/map.rs
struct feature_Map;
impl feature_Map {
    // ---- the third view in the picker --------------------------------------
    // /browse drew the row as "grid unless list", which cannot be extended by
    // appending: with the map chosen, grid would light up beside it. So the
    // row is restated here — three buttons, each lit by its own name — rather
    // than appended to. A fourth view does the same thing again, or /browse
    // grows a `browse_view_is(which)` predicate if that ever gets tiresome.

    fn browse_views() -> String {
        let view = browse_view_read();
        format!("{}{}{}",
                browse_view_button("grid".to_string(),
                                   view != "list" && view != "map"),
                browse_view_button("list".to_string(), view == "list"),
                browse_view_button("map".to_string(), view == "map"))
    }

    // the map's own button, drawn like its siblings; every other name goes on
    // down the chain to /browse's two. The data-ev is a literal string, not a
    // placeholder, so /sub-tool-cards' long-press can read it out of the
    // source.
    fn browse_view_button(which: String, on: bool) -> String {
        if which != "map" {
            return existing.browse_view_button(which, on);
        }
        let lit = if on { " browse-on" } else { "" };
        format!("<div class=\"browse-view{}\" data-ev=\"browse_map\" title=\"map\">{}</div>",
                lit, map_fold_svg())
    }

    // ---- choosing it --------------------------------------------------------
    // /browse's own two clicks are unchanged; this is the third, and it does
    // exactly what they do — write the view, and put the set back if a card
    // was open, because the picker is on screen over a card page too.

    fn update(state: String, event: String) -> String {
        let state = existing.update(state, event.clone());
        let e: serde_json::Value = serde_json::from_str(&event)
            .unwrap_or(serde_json::Value::Null);
        if e["type"].as_str().unwrap_or("") != "click" {
            return state;
        }
        if e["ev"].as_str().unwrap_or("") != "browse_map" {
            return state;
        }
        browse_view_write("map".to_string());
        if !browse_open_read().is_empty() {
            browse_open_write(String::new());
        }
        state
    }

    // ---- the surface --------------------------------------------------------
    // `browse_set_html` is the grid/list switch, so it is also the map/not-map
    // switch: one link, and /browse's own two views are untouched behind it.
    // Note this runs BEFORE the empty-set check, so a map with no pins is
    // still a map — which is right: an empty grid has nothing to show, an
    // empty map still shows where you are.

    fn browse_set_html(cards: &Vec<serde_json::Value>) -> String {
        if browse_view_read() != "map" {
            return existing.browse_set_html(cards);
        }
        map_surface_html(cards)
    }

    // Rust draws no map. It draws the ARGUMENT: one empty element carrying the
    // located cards as JSON, which the page half reads after every repaint.
    // The map itself lives outside #app and survives the repaint that replaces
    // this element — /keep's idiom for furniture a repaint must not destroy.
    fn map_surface_html(cards: &Vec<serde_json::Value>) -> String {
        let mut pins: Vec<serde_json::Value> = Vec::new();
        for c in cards.iter() {
            let at = card_place_of(c.to_string());
            if at.is_null() {
                continue;
            }
            pins.push(serde_json::json!({
                "id": c["id"].as_str().unwrap_or(""),
                "lat": at["lat"].as_f64().unwrap_or(0.0),
                "lon": at["lon"].as_f64().unwrap_or(0.0),
                "face": map_face_of(c),
                "initial": map_initial_of(c),
                "title": map_title_of(c)
            }));
        }
        let json = serde_json::Value::Array(pins).to_string();
        format!("<div id=\"mapData\" data-pins=\"{}\"></div>", card_esc(json))
    }

    // the card's own face, as the tile draws it: the first picture block, or
    // nothing and then the title's first character. The pin wears one or the
    // other, so a person is recognisable on the map at a glance.
    fn map_face_of(card: &serde_json::Value) -> String {
        let empty: Vec<serde_json::Value> = Vec::new();
        for b in card["blocks"].as_array().unwrap_or(&empty) {
            if b["kind"].as_str().unwrap_or("") == "picture" {
                return b["data"].as_str().unwrap_or("").to_string();
            }
        }
        String::new()
    }

    fn map_title_of(card: &serde_json::Value) -> String {
        let empty: Vec<serde_json::Value> = Vec::new();
        for b in card["blocks"].as_array().unwrap_or(&empty) {
            if b["kind"].as_str().unwrap_or("") == "title" {
                return b["text"].as_str().unwrap_or("").to_string();
            }
        }
        String::new()
    }

    fn map_initial_of(card: &serde_json::Value) -> String {
        // a card with no title (a post) pins with its owner's initial, as
        // /browse's row does (found by /posts, same day)
        let title = map_title_of(card);
        let from = if title.is_empty() {
            card["owner"].as_str().unwrap_or("").to_string()
        } else {
            title
        };
        from.chars().take(1).collect()
    }

    // ---- the glyph -----------------------------------------------------------
    // a folded map, drawn in currentColor (/glyphs). NOT a pin: the pins are
    // the things ON this view, and a screen carrying the same shape as both
    // its control and its content reads as a mistake — /browse made the same
    // call when it refused four squares for the tool that holds the grid.

    fn map_fold_svg() -> String {
        String::from(concat!(
            "<svg class=\"icon-svg\" viewBox=\"0 0 24 24\" aria-hidden=\"true\">",
            "<path d=\"M9 4 3 6v14l6-2 6 2 6-2V4l-6 2z\" fill=\"none\" stroke=\"currentColor\" stroke-width=\"2.2\" stroke-linejoin=\"round\"/>",
            "<path d=\"M9 4v14M15 6v14\" fill=\"none\" stroke=\"currentColor\" stroke-width=\"2.2\" stroke-linejoin=\"round\"/>",
            "</svg>"))
    }
}
